evm/src/propeller-router/PropellerRouter.sol

```diff
@@ -0,0 +1,634 @@
+
+        // For native ETH, assume funds already in our router. Else, transfer.
+        if (tokenIn != address(0)) {
+            IERC20(tokenIn).safeTransferFrom(msg.sender, address(this), givenAmount);
+        }
+
+
+        while (swaps.length > 0) {
+            (swap, swaps) = swaps.next();
+            exchange = swap.exchange();
+            calculatedAmount =
+                _executeSwap(exchange, calculatedAmount, swap.protocolData());
+        }
+        if (calculatedAmount < minUserAmount) {
+            revert NegativeSlippage(calculatedAmount, minUserAmount);
+        }
+    }
+
+    /**
+     * @dev Executes a sequence of exact out swaps, by first quoting
+     *  backwards and then executing with corrected amounts a
+     *  sequential exactIn swap
+     *
+     * This method checks that the user spends no more than maxUserAmount of sellToken
+     *  Note: All used executors must implement ISwapQuoter, for this
+     *  method to work correctly.
+     */
+    function sequentialExactOut(
+        uint256 givenAmount,
+        uint256 maxUserAmount,
+        bytes[] calldata swaps
+    )
+        external
+        override
+        onlyRole(EXECUTOR_ROLE)
+        withSwapContext
+        returns (uint256 calculatedAmount)
+    {
+        // Idea: On v2, reserve 14 bytes for calculatedAmount and replace them here
+        //  to save some quotes, if these 14 bytes are all zero the swap call won't
+        //  recalculate the quote else, it will simply execute with the calculatedAmount
+        //  that was passed along.
+        // TODO: Check if replacing 14 bytes (requires us to construct new swap in mempy)
+        //  amortises the repeated quoting (quoting again costs at least 500 gas, from
+        //  a quick calculation it should amortise)
+        bytes calldata swap;
+        uint256[] memory amounts = new uint256[](swaps.length + 1);
+        amounts[swaps.length] = givenAmount;
+
+        // backwards pass to get correct in amount
+        for (uint256 i = swaps.length; i > 0; i--) {
+            swap = swaps[i - 1];
+            amounts[i - 1] =
+                _quoteSwap(swap.exchange(), amounts[i], swap.protocolData());
+        }
+        for (uint8 i = 0; i < swaps.length; i++) {
+            swap = swaps[i];
+            _executeSwap(swap.exchange(), amounts[i + 1], swap.protocolData());
+        }
+        calculatedAmount = amounts[0];
+
+        if (calculatedAmount > maxUserAmount) {
+            revert NegativeSlippage(calculatedAmount, maxUserAmount);
+        }
+    }
+
+    /**
+     * @dev Executes a swap graph with internal splits token amount
+     *  splits, checking that the user gets more than minUserAmount of buyToken.
+     *
+     *  Assumes the swaps in swaps_ already contain any required token
+     *  addresses.
+     */
+    function splitExactIn(
+        uint256 amountIn,
+        address tokenIn,
+        uint256 minUserAmount,
+        SplitSwapExactInParameters calldata parameters
```

Comment:
what does it mean for this to be `parameters` and not bytes? 🤔

Reply:
It just automatically parses it into `SplitSwapExactInParameters`. We can change this - please don't take this PR too much as the ground truth. It's really meant for a sketch and nothing more - a few things will have to be polished here, but it was far too much to do during cooldown, so we will need to think at every step of implementation whether this choice is still the right choice. I hope that makes sense